In [1]:
import pandas as pd
import sqlalchemy as sa
import psycopg2 as ps
from sqlalchemy import create_engine

In [2]:
%load_ext sql
%sql postgresql://postgres:lingga28@localhost:2828/datacamp
conn = create_engine('postgresql://postgres:lingga28@localhost/datacamp')

# 1. Groups of customers
### Exercises
Use the CUBE operator to extract the content of a pivot table from the database. Create a table with the total number of male and female customers from each country.

### Instruction
Create a table with the total number of customers, of all female and male customers, of the number of customers for each country and the number of men and women from each country.

In [3]:
%%sql

SELECT gender, -- Extract information of a pivot table of gender and country for the number of customers
	   country,
	   COUNT(*)
FROM customers
GROUP BY CUBE (gender, country)
ORDER BY country
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(None, 'Austria', 4),
 ('male', 'Austria', 1),
 ('female', 'Austria', 3),
 (None, 'Belgium', 6),
 ('male', 'Belgium', 3),
 ('female', 'Belgium', 3),
 ('female', 'Denmark', 4),
 (None, 'Denmark', 7),
 ('male', 'Denmark', 3),
 ('female', 'France', 5)]

# 2. Categories of movies
### Exercises
Give an overview on the movies available on MovieNow. List the number of movies for different genres and release years.

### task 1
### Instruction
List the number of movies for different genres and the year of release on all aggregation levels by using the CUBE operator.

In [4]:
%%sql

SELECT genre,
       year_of_release,
       COUNT(*)
FROM movies
GROUP BY CUBE(genre, year_of_release)
ORDER BY year_of_release
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Science Fiction & Fantasy', 2001, 2),
 ('Comedy', 2001, 2),
 (None, 2001, 6),
 ('Drama', 2001, 2),
 ('Science Fiction & Fantasy', 2002, 2),
 (None, 2002, 7),
 ('Drama', 2002, 2),
 ('Comedy', 2002, 3),
 ('Drama', 2003, 3),
 ('Comedy', 2003, 2)]

### task 2
### Instruction
Question

Which statement is NOT correct about the result table?

### Possible Answers
- A. From all genres (ignoring the year of release) there are most movies in the category Drama.
- B. In total there are 71 movies available on MovieNow.
- C. The year of release with most movies is 2014.
- D. From 2002 there are 2 dramas available on MovieNow.

Answer: C

# 3. Analyzing average ratings
### Exercises
Prepare a table for a report about the national preferences of the customers from MovieNow comparing the average rating of movies across countries and genres.

### task 1
### Instruction
Augment the records of movie rentals with information about movies and customers, in this order. Use the first letter of the table names as alias.

In [5]:
%%sql

-- Augment the records of movie rentals with information about movies and customers
SELECT *
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 8, 'Waking Up in Reno', 'Comedy', 91, 2002, Decimal('2.59'), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12)),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79'), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39'), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13)),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59'), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19)),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69'), 104, 'Dimitri Zambrano Morales', 'Spain', 'male', datetime.date(1992, 6, 21), datetime.date(2017, 4, 12)),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79'), 50, 'Csordás Ármin', 'Hungary', 'male', datetime.date(1986, 11, 30), datetime.date(2018, 4, 23)),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 21, 'The Return of the King', 'Science Fiction & Fantasy', 200, 2003, Decimal('1.99'), 52, 'Szepessi Marton', 'Hungary', 'male', datetime.date(1983, 3, 14), datetime.date(2018, 3, 24)),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79'), 73, 'Veneranda Milani', 'Italy', 'female', datetime.date(1983, 1, 15), datetime.date(2018, 1, 15)),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 2, 'Swordfish', 'Drama', 99, 2001, Decimal('2.19'), 78, 'Doroteo Napolitano', 'Italy', 'male', datetime.date(1992, 6, 1), datetime.date(2017, 7, 27)),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09'), 121, 'Annie R. Barnett', 'USA', 'female', datetime.date(1987, 3, 2), datetime.date(2017, 9, 13))]

### task 2
### Instruction
Calculate the average rating for each country.

In [6]:
%%sql

-- Calculate the average rating for each country
SELECT 
	c.country,
    AVG(r.rating)
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY country;

 * postgresql://postgres:***@localhost:2828/datacamp
12 rows affected.


[(None, Decimal('7.5000000000000000')),
 ('Spain', Decimal('7.6415094339622642')),
 ('Great Britan', Decimal('7.6538461538461538')),
 ('Austria', Decimal('6.8000000000000000')),
 ('Poland', Decimal('8.1212121212121212')),
 ('Italy', Decimal('8.1379310344827586')),
 ('Slovenia', Decimal('8.4318181818181818')),
 ('Hungary', Decimal('7.6285714285714286')),
 ('Denmark', Decimal('7.8888888888888889')),
 ('Belgium', Decimal('8.9000000000000000')),
 ('France', Decimal('7.7187500000000000')),
 ('USA', Decimal('8.0000000000000000'))]

### task 3
### Instruction
Calculate the average rating for all aggregation levels of country and genre.

In [7]:
%%sql

SELECT 
	c.country, 
	m.genre, 
	AVG(r.rating) AS avg_rating -- Calculate the average rating 
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY CUBE(country, genre) -- For all aggregation levels of country and genre
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(None, None, Decimal('7.9390243902439024')),
 ('France', 'Mystery & Suspense', None),
 ('Slovenia', 'Action & Adventure', None),
 ('Spain', 'Animation', None),
 ('Poland', 'Comedy', Decimal('8.2000000000000000')),
 ('Denmark', 'Drama', Decimal('7.6000000000000000')),
 ('Italy', 'Science Fiction & Fantasy', Decimal('7.6666666666666667')),
 ('France', 'Science Fiction & Fantasy', Decimal('7.6666666666666667')),
 (None, 'Drama', Decimal('6.0000000000000000')),
 ('USA', 'Mystery & Suspense', Decimal('3.0000000000000000'))]

### task 4
### Instruction
Question

What is the average rating over all records, rounded to two digits?

### Possible Answers
- A. 7.94
- B. null
- C. 8.80
- D. 7.86

Answer: A

# 4. Number of customers
### Exercises
You have to give an overview of the number of customers for a presentation.

### Instructions
- Generate a table with the total number of customers, the number of customers for each country, and the number of female and male customers for each country.
- Order the result by country and gender.

In [8]:
%%sql

-- Count the total number of customers, the number of customers for each country, and the number of female and male customers for each country
SELECT country,
       gender,
	   COUNT(*)
FROM customers
GROUP BY ROLLUP (country, gender)
ORDER BY country, gender -- Order the result by country and gender
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Austria', 'female', 3),
 ('Austria', 'male', 1),
 ('Austria', None, 4),
 ('Belgium', 'female', 3),
 ('Belgium', 'male', 3),
 ('Belgium', None, 6),
 ('Denmark', 'female', 4),
 ('Denmark', 'male', 3),
 ('Denmark', None, 7),
 ('France', 'female', 5)]

# 5. Analyzing preferences of genres across countries
### Exercises
You are asked to study the preferences of genres across countries. Are there particular genres which are more popular in specific countries? Evaluate the preferences of customers by averaging their ratings and counting the number of movies rented from each genre.

### task 1
### Instruction
Augment the renting records with information about movies and customers.

In [9]:
%%sql

-- Join the tables
SELECT *
FROM renting AS r
LEFT JOIN movies AS m
ON r.movie_id = m.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 8, 'Waking Up in Reno', 'Comedy', 91, 2002, Decimal('2.59'), 41, 'Zara Mitchell', 'Great Britan', 'female', datetime.date(1994, 7, 8), datetime.date(2017, 6, 12)),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79'), 10, 'Arnout Veenhuis', 'Belgium', 'male', datetime.date(1984, 7, 26), datetime.date(2017, 1, 28)),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39'), 108, 'Saúl Tafoya Meraz', 'Spain', 'male', datetime.date(1992, 5, 15), datetime.date(2017, 3, 13)),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59'), 39, 'Amy Haynes', 'Great Britan', 'female', datetime.date(1975, 7, 28), datetime.date(2018, 1, 19)),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69'), 104, 'Dimitri Zambrano Morales', 'Spain', 'male', datetime.date(1992, 6, 21), datetime.date(2017, 4, 12)),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79'), 50, 'Csordás Ármin', 'Hungary', 'male', datetime.date(1986, 11, 30), datetime.date(2018, 4, 23)),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 21, 'The Return of the King', 'Science Fiction & Fantasy', 200, 2003, Decimal('1.99'), 52, 'Szepessi Marton', 'Hungary', 'male', datetime.date(1983, 3, 14), datetime.date(2018, 3, 24)),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79'), 73, 'Veneranda Milani', 'Italy', 'female', datetime.date(1983, 1, 15), datetime.date(2018, 1, 15)),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 2, 'Swordfish', 'Drama', 99, 2001, Decimal('2.19'), 78, 'Doroteo Napolitano', 'Italy', 'male', datetime.date(1992, 6, 1), datetime.date(2017, 7, 27)),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09'), 121, 'Annie R. Barnett', 'USA', 'female', datetime.date(1987, 3, 2), datetime.date(2017, 9, 13))]

### task 2
### Instruction
Calculate the average ratings and the number of ratings for each country and each genre. Include the columns country and genre in the SELECT clause.

In [10]:
%%sql

SELECT 
	c.country, -- Select country
	m.genre, -- Select genre
	AVG(r.rating), -- Average ratings
	COUNT(*)  -- Count number of movie rentals
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY c.country, m.genre -- Aggregate for each country and each genre
ORDER BY c.country, m.genre
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Austria', 'Comedy', Decimal('8.0000000000000000'), 1),
 ('Austria', 'Drama', Decimal('6.0000000000000000'), 2),
 ('Austria', 'Mystery & Suspense', None, 1),
 ('Austria', 'Science Fiction & Fantasy', Decimal('6.6666666666666667'), 3),
 ('Belgium', 'Comedy', None, 1),
 ('Belgium', 'Drama', Decimal('9.1666666666666667'), 15),
 ('Belgium', 'Mystery & Suspense', None, 1),
 ('Belgium', 'Science Fiction & Fantasy', Decimal('8.5000000000000000'), 7),
 ('Denmark', 'Drama', Decimal('7.6000000000000000'), 12),
 ('Denmark', 'Mystery & Suspense', Decimal('7.2500000000000000'), 4)]

### task 3
### Instruction
Finally, calculate the average ratings and the number of ratings for each country and genre, as well as an aggregation over all genres for each country and the overall average and total number.

In [11]:
%%sql

-- Group by each county and genre with OLAP extension
SELECT 
	c.country, 
	m.genre, 
	AVG(r.rating) AS avg_rating, 
	COUNT(*) AS num_rating
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY ROLLUP (c.country, m.genre)
ORDER BY c.country, m.genre
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Austria', 'Comedy', Decimal('8.0000000000000000'), 1),
 ('Austria', 'Drama', Decimal('6.0000000000000000'), 2),
 ('Austria', 'Mystery & Suspense', None, 1),
 ('Austria', 'Science Fiction & Fantasy', Decimal('6.6666666666666667'), 3),
 ('Austria', None, Decimal('6.8000000000000000'), 7),
 ('Belgium', 'Comedy', None, 1),
 ('Belgium', 'Drama', Decimal('9.1666666666666667'), 15),
 ('Belgium', 'Mystery & Suspense', None, 1),
 ('Belgium', 'Science Fiction & Fantasy', Decimal('8.5000000000000000'), 7),
 ('Belgium', None, Decimal('8.9000000000000000'), 24)]

# 6. Queries with GROUPING SETS
### Exercises
What question CANNOT be answered by the following query?

SELECT \
  r.customer_id, \
  m.genre, \
  AVG(r.rating), \
  COUNT(*)\
FROM renting AS r\
LEFT JOIN movies AS m\
ON r.movie_id = m.movie_id\
GROUP BY GROUPING SETS ((r.customer_id, m.genre), (r.customer_id), ());

### Possible Answers
- A. How many movies were watched by each customer?
- B. What is the average rating for each genre?
- C. What is the average rating of customer 75 for movies of the Comedy genre?
- D. What is the overall average rating for all movies from all customers?

Answer: B

# 7. Exploring nationality and gender of actors
### Exercises
For each movie in the database, the three most important actors are identified and listed in the table actors. This table includes the nationality and gender of the actors. We are interested in how much diversity there is in the nationalities of the actors and how many actors and actresses are in the list.

### Instruction
Count the number of actors in the table actors from each country, the number of male and female actors and the total number of actors.

In [12]:
%%sql

SELECT 
	nationality, -- Select nationality of the actors
    gender, -- Select gender of the actors
    COUNT(*) -- Count the number of actors
FROM actors
GROUP BY GROUPING SETS ((nationality), (gender), ()); -- Use the correct GROUPING SETS operation

 * postgresql://postgres:***@localhost:2828/datacamp
20 rows affected.


[(None, None, 145),
 ('Somalia', None, 1),
 (None, None, 2),
 ('Argentina', None, 1),
 ('Spain', None, 3),
 ('Italy', None, 1),
 ('Puerto Rico', None, 1),
 ('Iran', None, 1),
 ('Northern Ireland', None, 2),
 ('USA', None, 91),
 ('Israel', None, 1),
 ('Netherlands', None, 1),
 ('South Africa', None, 1),
 ('Austria', None, 1),
 ('Australia', None, 8),
 ('Ireland', None, 2),
 ('Canada', None, 4),
 ('British', None, 24),
 (None, 'male', 90),
 (None, 'female', 55)]

# 8. Exploring rating by country and gender
Now you will investigate the average rating of customers aggregated by country and gender.

### task 1
### Instruction
Select the columns country, gender, and rating and use the correct join to combine the table renting with customer.

In [13]:
%%sql

SELECT 
	c.country, -- Select country, gender and rating
    c.gender,
    r.rating
FROM renting AS r
LEFT JOIN customers AS c -- Use the correct join
ON r.customer_id = c.customer_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Great Britan', 'female', None),
 ('Belgium', 'male', 10),
 ('Spain', 'male', 4),
 ('Great Britan', 'female', 8),
 ('Spain', 'male', 7),
 ('Hungary', 'male', 7),
 ('Hungary', 'male', None),
 ('Italy', 'female', 10),
 ('Italy', 'male', None),
 ('USA', 'female', None)]

### task 2
### Instruction
Use GROUP BY to calculate the average rating over country and gender. Order the table by country and gender.

In [14]:
%%sql

SELECT 
	c.country, 
    c.gender,
	AVG(rating) -- Calculate average rating
FROM renting AS r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY country, gender -- Order and group by country and gender
ORDER BY country, gender
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Austria', 'female', Decimal('7.0000000000000000')),
 ('Austria', 'male', Decimal('6.0000000000000000')),
 ('Belgium', 'female', Decimal('9.1250000000000000')),
 ('Belgium', 'male', Decimal('8.0000000000000000')),
 ('Denmark', 'female', Decimal('8.4444444444444444')),
 ('Denmark', 'male', Decimal('7.3333333333333333')),
 ('France', 'female', Decimal('8.0000000000000000')),
 ('France', 'male', Decimal('7.5714285714285714')),
 ('Great Britan', 'female', Decimal('7.2727272727272727')),
 ('Great Britan', 'male', Decimal('7.9333333333333333'))]

### task 3
### Instruction
Now, use GROUPING SETS to get the same result, i.e. the average rating over country and gender.

In [15]:
%%sql

SELECT 
	c.country, 
    c.gender,
	AVG(r.rating)
FROM renting AS r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
GROUP BY GROUPING SETS ((country, gender)) -- Group by country and gender with GROUPING SETS
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Austria', 'male', Decimal('6.0000000000000000')),
 ('France', 'female', Decimal('8.0000000000000000')),
 ('Hungary', 'female', Decimal('7.2857142857142857')),
 (None, None, Decimal('7.5000000000000000')),
 ('Spain', 'female', Decimal('7.6129032258064516')),
 ('Belgium', 'male', Decimal('8.0000000000000000')),
 ('USA', 'male', Decimal('7.5000000000000000')),
 ('Denmark', 'female', Decimal('8.4444444444444444')),
 ('Austria', 'female', Decimal('7.0000000000000000')),
 ('Slovenia', 'male', Decimal('8.0909090909090909'))]

### task 4
### Instruction
Report all information that is included in a pivot table for country and gender in one SQL table.

In [16]:
%%sql

SELECT 
	c.country, 
    c.gender,
	AVG(r.rating)
FROM renting AS r
LEFT JOIN customers AS c
ON r.customer_id = c.customer_id
-- Report all info from a Pivot table for country and gender
GROUP BY GROUPING SETS ((country, gender), (country), (gender), ())
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(None, None, Decimal('7.9390243902439024')),
 ('Austria', 'male', Decimal('6.0000000000000000')),
 ('France', 'female', Decimal('8.0000000000000000')),
 ('Hungary', 'female', Decimal('7.2857142857142857')),
 (None, None, Decimal('7.5000000000000000')),
 ('Spain', 'female', Decimal('7.6129032258064516')),
 ('Belgium', 'male', Decimal('8.0000000000000000')),
 ('USA', 'male', Decimal('7.5000000000000000')),
 ('Denmark', 'female', Decimal('8.4444444444444444')),
 ('Austria', 'female', Decimal('7.0000000000000000'))]

# 9. Customer preference for genres
### Exercises
You just saw that customers have no clear preference for more recent movies over older ones. Now the management considers investing money in movies of the best rated genres.

### task 1
### Instruction
Augment the records of movie rentals with information about movies. Use the first letter of the table as alias.

In [17]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN movies AS m -- Augment the table with information about movies
ON r.movie_id = m.movie_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 8, 'Waking Up in Reno', 'Comedy', 91, 2002, Decimal('2.59')),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79')),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39')),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59')),
 (5, 104, 15, 7, datetime.date(2019, 3, 18), 15, "Something's Gotta Give", 'Comedy', 128, 2003, Decimal('1.69')),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79')),
 (7, 52, 21, None, datetime.date(2018, 11, 10), 21, 'The Return of the King', 'Science Fiction & Fantasy', 200, 2003, Decimal('1.99')),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79')),
 (9, 78, 2, None, datetime.date(2017, 9, 3), 2, 'Swordfish', 'Drama', 99, 2001, Decimal('2.19')),
 (10, 121, 43, None, datetime.date(2017, 11, 8), 43, 'The Invasion', 'Mystery & Suspense', 99, 2007, Decimal('2.09'))]

### task 2
### Instruction
Select records of movies with at least 4 ratings, starting from 2018-04-01.

In [18]:
%%sql

SELECT *
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( -- Select records of movies with at least 4 ratings
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01' -- Select records of movie rentals since 2018-04-01
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(3, 108, 45, 4, datetime.date(2018, 6, 8), 45, 'Burn After Reading', 'Drama', 96, 2008, Decimal('2.39')),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 66, 'The Hunger Games', 'Drama', 142, 2012, Decimal('1.59')),
 (6, 50, 71, 7, datetime.date(2018, 10, 9), 71, 'Rio 2', 'Animation', 101, 2014, Decimal('1.79')),
 (8, 73, 65, 10, datetime.date(2018, 6, 5), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79')),
 (12, 52, 65, 10, datetime.date(2018, 6, 29), 65, 'Ghost Rider: Spirit of Vengeance', 'Action & Adventure', 96, 2012, Decimal('1.79')),
 (14, 8, 29, None, datetime.date(2018, 8, 3), 29, 'Two for the Money', 'Drama', 122, 2005, Decimal('2.79')),
 (17, 22, 46, 10, datetime.date(2019, 2, 16), 46, 'Sleepwalking', 'Drama', 101, 2008, Decimal('1.79')),
 (18, 36, 39, 10, datetime.date(2019, 3, 20), 39, "I'm Not There", 'Drama', 135, 2007, Decimal('1.79')),
 (19, 99, 12, None, datetime.date(2018, 9, 16), 12, 'The Two Towers', 'Science Fiction & Fantasy', 179, 2002, Decimal('2.39')),
 (20, 73, 32, None, datetime.date(2018, 9, 1), 32, 'A Good Woman', 'Drama', 93, 2006, Decimal('2.89'))]

### task 3
### Instruction
For each genre, calculate the average rating (use the alias avg_rating), the number of ratings (use the alias n_rating), the number of movie rentals (use the alias n_rentals), and the number of distinct movies (use the alias n_movies).

In [19]:
%%sql

SELECT m.genre, -- For each genre, calculate:
	   AVG(r.rating) AS avg_rating, -- The average rating and use the alias avg_rating
	   COUNT(r.rating) AS n_rating, -- The number of ratings and use the alias n_rating
	   COUNT(*) AS n_rentals,     -- The number of movie rentals and use the alias n_rentals
	   COUNT(DISTINCT m.movie_id) AS n_movies -- The number of distinct movies and use the alias n_movies
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3)
AND r.date_renting >= '2018-01-01'
GROUP BY m.genre;

 * postgresql://postgres:***@localhost:2828/datacamp
8 rows affected.


[('Action & Adventure', Decimal('8.7142857142857143'), 7, 9, 2),
 ('Animation', Decimal('7.8333333333333333'), 6, 10, 2),
 ('Art House & International', Decimal('8.5000000000000000'), 4, 5, 1),
 ('Comedy', Decimal('7.9500000000000000'), 20, 31, 5),
 ('Drama', Decimal('7.7482517482517483'), 143, 245, 34),
 ('Mystery & Suspense', Decimal('7.4285714285714286'), 7, 19, 3),
 ('Other', Decimal('8.4285714285714286'), 7, 16, 2),
 ('Science Fiction & Fantasy', Decimal('8.2765957446808511'), 47, 70, 10)]

### task 4
### Instruction
Order the table by decreasing average rating.

In [20]:
%%sql

SELECT genre,
	   AVG(rating) AS avg_rating,
	   COUNT(rating) AS n_rating,
       COUNT(*) AS n_rentals,     
	   COUNT(DISTINCT m.movie_id) AS n_movies 
FROM renting AS r
LEFT JOIN movies AS m
ON m.movie_id = r.movie_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 3 )
AND r.date_renting >= '2018-01-01'
GROUP BY genre
ORDER BY avg_rating DESC; -- Order the table by decreasing average rating

 * postgresql://postgres:***@localhost:2828/datacamp
8 rows affected.


[('Action & Adventure', Decimal('8.7142857142857143'), 7, 9, 2),
 ('Art House & International', Decimal('8.5000000000000000'), 4, 5, 1),
 ('Other', Decimal('8.4285714285714286'), 7, 16, 2),
 ('Science Fiction & Fantasy', Decimal('8.2765957446808511'), 47, 70, 10),
 ('Comedy', Decimal('7.9500000000000000'), 20, 31, 5),
 ('Animation', Decimal('7.8333333333333333'), 6, 10, 2),
 ('Drama', Decimal('7.7482517482517483'), 143, 245, 34),
 ('Mystery & Suspense', Decimal('7.4285714285714286'), 7, 19, 3)]

# 10. Customer preference for actors
The last aspect you have to analyze are customer preferences for certain actors.

### task 1
### Instruction
Join the tables.

In [21]:
%%sql

-- Join the tables
SELECT *
FROM renting AS r
LEFT JOIN actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN actors AS a
ON ai.actor_id = a.actor_id
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[(1, 41, 8, None, datetime.date(2018, 10, 9), 160, 8, 107, 107, 'Patrick Swayze', 1952, 'USA', 'male'),
 (1, 41, 8, None, datetime.date(2018, 10, 9), 152, 8, 103, 103, 'Natasha Richardson', 1963, 'British', 'female'),
 (1, 41, 8, None, datetime.date(2018, 10, 9), 28, 8, 23, 23, 'Charlize Theron', 1975, 'South Africa', 'female'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 172, 29, 117, 117, 'Rene Russo', 1954, 'USA', 'female'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 143, 29, 96, 96, 'Matthew McConaughey', 1969, 'USA', 'male'),
 (2, 10, 29, 10, datetime.date(2017, 3, 1), 5, 29, 3, 3, 'Al Pacino', 1940, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 118, 45, 74, 74, 'John Malkovich', 1953, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 80, 45, 49, 49, 'George Clooney', 1961, 'USA', 'male'),
 (3, 108, 45, 4, datetime.date(2018, 6, 8), 74, 45, 45, 45, 'Frances McDormand', 1957, 'USA', 'female'),
 (4, 39, 66, 8, datetime.date(2018, 10, 22), 130, 66, 86, 86, 'Liam Hemsworth', 1990, 'Australia', 'male')]

### task 2
### Instruction
For each combination of the actors' nationality and gender, calculate the average rating, the number of ratings, the number of movie rentals, and the number of actors.

In [22]:
%%sql

SELECT a.nationality,
       a.gender,
	   AVG(r.rating) AS avg_rating, -- The average rating
	   COUNT(r.rating) AS n_rating, -- The number of ratings
	   COUNT(*) AS n_rentals, -- The number of movie rentals
	   COUNT(DISTINCT a.actor_id) AS n_actors -- The number of actors
FROM renting AS r
LEFT JOIN actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >=4 )
AND r.date_renting >= '2018-04-01'
GROUP BY a.nationality, a.gender;

 * postgresql://postgres:***@localhost:2828/datacamp
19 rows affected.


[('Argentina', 'male', Decimal('8.5000000000000000'), 4, 5, 1),
 ('Australia', 'female', Decimal('8.6666666666666667'), 3, 5, 1),
 ('Australia', 'male', Decimal('7.4545454545454545'), 11, 17, 3),
 ('Austria', 'male', Decimal('8.5000000000000000'), 2, 6, 1),
 ('British', 'female', Decimal('7.8333333333333333'), 54, 78, 3),
 ('British', 'male', Decimal('8.1052631578947368'), 114, 175, 9),
 ('Canada', 'female', Decimal('8.0000000000000000'), 5, 5, 1),
 ('Canada', 'male', Decimal('8.4285714285714286'), 7, 13, 2),
 ('Ireland', 'male', Decimal('7.7142857142857143'), 7, 16, 2),
 ('Israel', 'female', Decimal('8.2500000000000000'), 4, 8, 1),
 ('Italy', 'female', Decimal('9.0000000000000000'), 4, 4, 1),
 ('Northern Ireland', 'male', Decimal('9.0000000000000000'), 4, 4, 1),
 ('Puerto Rico', 'male', Decimal('7.2000000000000000'), 5, 7, 1),
 ('Somalia', 'male', Decimal('7.0000000000000000'), 3, 5, 1),
 ('South Africa', 'female', Decimal('8.2142857142857143'), 14, 21, 1),
 ('Spain', 'female', Decimal('8.5000000000000000'), 4, 5, 1),
 ('Spain', 'male', Decimal('8.8571428571428571'), 7, 12, 2),
 ('USA', 'female', Decimal('7.5454545454545455'), 121, 212, 27),
 ('USA', 'male', Decimal('7.9132231404958678'), 242, 401, 42)]

### task 3
### Instruction
Provide results for all aggregation levels represented in a pivot table.

In [23]:
%%sql

SELECT a.nationality,
       a.gender,
	   AVG(r.rating) AS avg_rating,
	   COUNT(r.rating) AS n_rating,
	   COUNT(*) AS n_rentals,
	   COUNT(DISTINCT a.actor_id) AS n_actors
FROM renting AS r
LEFT JOIN actsin AS ai
ON ai.movie_id = r.movie_id
LEFT JOIN actors AS a
ON ai.actor_id = a.actor_id
WHERE r.movie_id IN ( 
	SELECT movie_id
	FROM renting
	GROUP BY movie_id
	HAVING COUNT(rating) >= 4)
AND r.date_renting >= '2018-04-01'
GROUP BY CUBE(a.nationality, a.gender) -- Provide results for all aggregation levels represented in a pivot table
LIMIT 10; --just an addition, so that the table is not elongated

 * postgresql://postgres:***@localhost:2828/datacamp
10 rows affected.


[('Argentina', 'male', Decimal('8.5000000000000000'), 4, 5, 1),
 ('Argentina', None, Decimal('8.5000000000000000'), 4, 5, 1),
 ('Australia', 'female', Decimal('8.6666666666666667'), 3, 5, 1),
 ('Australia', 'male', Decimal('7.4545454545454545'), 11, 17, 3),
 ('Australia', None, Decimal('7.7142857142857143'), 14, 22, 4),
 ('Austria', 'male', Decimal('8.5000000000000000'), 2, 6, 1),
 ('Austria', None, Decimal('8.5000000000000000'), 2, 6, 1),
 ('British', 'female', Decimal('7.8333333333333333'), 54, 78, 3),
 ('British', 'male', Decimal('8.1052631578947368'), 114, 175, 9),
 ('British', None, Decimal('8.0178571428571429'), 168, 253, 12)]